core/src/main/codegen/templates/Parser.jj

```diff
@@ -1735,44 +1742,104 @@ SqlNode SqlMerge() :
     [ [ <AS> ] alias = SimpleIdentifier() ]
     <USING> sourceTableRef = TableRef()
     <ON> condition = Expression(ExprContext.ACCEPT_SUB_QUERY)
+
     (
+        //Need look ahead here to determine if we have any update call whatsoever
         LOOKAHEAD(2)
```

Comment:
Basically, whenever the parser reaches a "choice point", it will look ahead some number of tokens, and take the first matching choice. A choice point is any point where you have to decide between one or more "paths" (aka expansions) that you could go down.

In this specific example, at the both of the choice points, we need to distinguish WHEN MATCHED from WHEN NOT MATCHED. I've updated the comments, lmk if it makes more sense